prototypes/ScheduledMerges.hs

```diff
@@ -506,25 +519,143 @@ assert p x = Exc.assert p (const x callStack)
+                   MergeCredit {spentCredits, unspentCredits} =
+    let suppliedCredits = spentCredits + unspentCredits
+     in spentCredits    >= 0
+     -- unspentCredits could legitimately be negative, though that does not
+     -- happen in this prototype
+     && suppliedCredits >= 0
+     && suppliedCredits <= totalDebt
+
+mergeDebtLeft :: HasCallStack => MergeDebt -> MergeCredit -> Debt
+mergeDebtLeft MergeDebt {totalDebt}
+              MergeCredit {spentCredits, unspentCredits} =
+    let suppliedCredits = spentCredits + unspentCredits
+     in assert (suppliedCredits <= totalDebt)
+               (totalDebt - suppliedCredits)
+
+-- | As credits are paid, debt is reduced in batches when sufficient credits
+-- have accumulated.
+data MergeDebtPaydown =
+    -- | This remaining merge debt is fully paid off, potentially with
+    -- leftovers.
+    MergeDebtDischarged !Debt !Credit
+
+    -- | Credits were paid, but not enough for merge debt to be reduced by some
+    -- batches of merging work.
+  | MergeDebtPaydownCredited !MergeCredit
+
+    -- | Enough credits were paid to reduce merge debt by performing some
+    -- batches of merging work.
+  | MergeDebtPaydownPerform !Debt !MergeCredit
+  deriving stock Show
+
+-- | Pay credits to merge debt, which might trigger performing some merge work
+-- in batches. See 'MergeDebtPaydown'.
+--
+paydownMergeDebt :: MergeDebt -> MergeCredit -> Credit -> MergeDebtPaydown
+paydownMergeDebt MergeDebt {totalDebt}
+                 MergeCredit {spentCredits, unspentCredits}
+                 c
+  | assert (c >= 0) False = undefined
+
+  | let !suppliedCredits' = suppliedCredits + c
+  , suppliedCredits' >= totalDebt
+  , let !leftover = suppliedCredits' - totalDebt
+        !perform  = c - leftover
+  = assert (perform >= 0 && leftover >= 0) $
+    assert (c == perform + leftover) $
+    assert (suppliedCredits + perform == totalDebt) $
+    MergeDebtDischarged perform leftover
+
+  | let !unspentCredits' = unspentCredits + c
+  , unspentCredits' >= mergeBatchSize
+  , let (!b, !r)         = divMod unspentCredits' mergeBatchSize
+        !perform         = b * mergeBatchSize
+        spentCredits'    = spentCredits    + perform
+        unspentCredits'' = unspentCredits' - perform
+        suppliedCredits' = spentCredits' + unspentCredits''
+  = assert (unspentCredits'' == r)
+    assert (suppliedCredits + c == suppliedCredits') $
+    assert (suppliedCredits' < totalDebt) $
+    MergeDebtPaydownPerform
+      perform
+      MergeCredit {
+        spentCredits   = spentCredits',
+        unspentCredits = unspentCredits''
+      }
+
+  | otherwise
+  , let unspentCredits'  = unspentCredits + c
+        suppliedCredits' = spentCredits + unspentCredits'
+  = assert (suppliedCredits + c == suppliedCredits') $
+    assert (suppliedCredits' < totalDebt) $
+    MergeDebtPaydownCredited
+      MergeCredit {
+        spentCredits,
+        unspentCredits = unspentCredits'
+      }
```

Comment:
This gets a bit noisy with all the assertions and guards.

Still good to have assertions, but maybe if we move most of this code into a helper function in the where-clause we can `assert (c >= 0)` at the top level without abusing a guard for it and also define `suppliedCredits' = suppliedCredits + c` just once and then assert it is equal to the alternative definitions in each branch.

Reply:
Ok I've refactored this so that the assertions are out of line and common definitions are factored out into the where clause. The main body is now a lot shorter (fits on a page).